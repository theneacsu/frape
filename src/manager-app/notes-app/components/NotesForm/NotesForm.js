import React, { Component } from 'react'
import styles from './NotesForm.module.css'

class NotesForm extends Component {
  state = {
    title: this.props.note ? this.props.note.title : '',
    body: this.props.note ? this.props.note.body : '',
    error: undefined
  }

  onInputChange = e => {
    const title = e.target.value
    this.setState(() => ({ title }))
  }

  onTextareaChange = e => {
    const body = e.target.value
    this.setState(() => ({ body }))
  }

  onFormSubmit = e => {
    e.preventDefault()

    if (this.state.title.trim().length > 2) {
      const data = {
        body: this.state.body,
        title: this.state.title
      }
      this.props.onSubmit(data)
    } else {
      const error = 'The title must have at least 3 characters.'
      this.setState(() => ({ error }))
    }
  }

  onButtonRemove = () => {
    this.props.onRemove()
  }

  render() {
    return (
      <div className={styles.wrapperDiv}>
        <div className="container">
          <form className={styles.form} onSubmit={this.onFormSubmit}>
            <input
              className="input"
              type="text"
              placeholder={this.props.titlePlaceholder}
              value={this.state.title}
              onChange={this.onInputChange}
            />
            {this.state.error && <p className="error">{this.state.error}</p>}
            <textarea
              className={styles.textarea}
              placeholder={this.props.bodyPlaceholder}
              value={this.state.body}
              onChange={this.onTextareaChange}
            />
            {this.props.purpose === 'edit' ? (
              <button
                type="submit"
                className={[styles.button, styles.save].join(' ')}
              >
                Save
              </button>
            ) : (
              <button
                type="submit"
                className={[styles.button, styles.save].join(' ')}
              >
                Add
              </button>
            )}
          </form>
          {this.props.purpose === 'edit' && (
            <button
              onClick={this.onButtonRemove}
              className={[styles.button, styles.delete].join(' ')}
            >
              Delete
            </button>
          )}
        </div>
      </div>
    )
  }
}

export default NotesForm
