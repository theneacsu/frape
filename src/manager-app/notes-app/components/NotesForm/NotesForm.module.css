.wrapperDiv {
  background-color: rgb(250, 250, 190);
  flex-grow: 1;
}

.div {
  display: flex;
  flex-direction: column;
}

.form {
  display: flex;
  padding-top: 2rem;
  flex-direction: column;
}

.textarea {
  margin-top: 1.5rem;
  outline: none;
  border: none;
  padding: 1.5rem;
  margin-bottom: 0.4rem;
  color: #47667c;
  border-radius: 1rem;
  width: 100%;
  display: block;
  resize: none;
  height: 40rem;
}

.textarea::placeholder {
  color: #ccc;
}

.button {
  border: none;
  border-radius: 1rem;
  padding: 1.5rem;
  color: white;
  font-weight: bold;
  width: 100%;
  margin-bottom: 1.5rem;
}

.save {
  margin-top: 1.5rem;
  background-color: #aa76f8;
}

.delete {
  background-color: #dd4b39;
}
