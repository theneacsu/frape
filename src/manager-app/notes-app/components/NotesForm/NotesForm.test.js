import React from 'react'
import { shallow } from 'enzyme'
import NotesForm from './NotesForm'

let purpose,
  titlePlaceholder,
  bodyPlaceholder,
  onSubmitSpy,
  noteData,
  onRemoveSpy,
  history

beforeEach(() => {
  noteData = {
    text: 'Eat fruits',
    body: 'A lot of them'
  }
  titlePlaceholder = 'a title'
  bodyPlaceholder = 'a body'
  purpose = 'add'
  onSubmitSpy = jest.fn()
  onRemoveSpy = jest.fn()
  history = {
    push: jest.fn()
  }
})

test('Should render the NotesForm component when adding a new note correctly', () => {
  const wrapper = shallow(
    <NotesForm
      purpose={purpose}
      titlePlaceholde={titlePlaceholder}
      bodyPlaceholder={bodyPlaceholder}
      onSubmit={onSubmitSpy}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

test('Should render the NotesForm component when editing a note correctly', () => {
  purpose = 'edit'
  const wrapper = shallow(
    <NotesForm
      purpose={purpose}
      note={noteData}
      onSubmit={onSubmitSpy}
      onRemove={onRemoveSpy}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

test('Should set state correctly when input change', () => {
  const value = 'headphones'
  const wrapper = shallow(
    <NotesForm
      purpose={purpose}
      titlePlaceholde={titlePlaceholder}
      bodyPlaceholder={bodyPlaceholder}
      onSubmit={onSubmitSpy}
    />
  )
  wrapper
    .find('input')
    .at(0)
    .simulate('change', {
      target: {
        value
      }
    })
  expect(wrapper.state('title')).toBe(value)
})

test('Should set state correctly when textarea change', () => {
  const value = 'go home'
  const wrapper = shallow(
    <NotesForm
      purpose={purpose}
      titlePlaceholde={titlePlaceholder}
      bodyPlaceholder={bodyPlaceholder}
      onSubmit={onSubmitSpy}
    />
  )
  wrapper
    .find('textarea')
    .at(0)
    .simulate('change', {
      target: {
        value
      }
    })
  expect(wrapper.state('body')).toBe(value)
})

test('Should set state correctly when invalid data is submited', () => {
  const wrapper = shallow(
    <NotesForm
      purpose={purpose}
      titlePlaceholde={titlePlaceholder}
      bodyPlaceholder={bodyPlaceholder}
      onSubmit={onSubmitSpy}
    />
  )
  wrapper.setState({ title: '' })
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  })
  expect(wrapper.state('error').length).toBeGreaterThan(0)
})
