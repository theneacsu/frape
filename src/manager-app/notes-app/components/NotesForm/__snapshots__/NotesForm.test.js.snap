// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the NotesForm component when adding a new note correctly 1`] = `
<div
  className="wrapperDiv"
>
  <div
    className="container"
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        className="input"
        onChange={[Function]}
        type="text"
        value=""
      />
      <textarea
        className="textarea"
        onChange={[Function]}
        placeholder="a body"
        value=""
      />
      <button
        className="button save"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</div>
`;

exports[`Should render the NotesForm component when editing a note correctly 1`] = `
<div
  className="wrapperDiv"
>
  <div
    className="container"
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        className="input"
        onChange={[Function]}
        type="text"
      />
      <textarea
        className="textarea"
        onChange={[Function]}
        value="A lot of them"
      />
      <button
        className="button save"
        type="submit"
      >
        Save
      </button>
    </form>
    <button
      className="button delete"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
